agent/src/util/device_plugin_service.rs

```diff
@@ -403,100 +809,337 @@ impl InstanceDevicePlugin {
+            node_name,
+            instance_name,
+            instance_namespace,
+            capacity,
+            kube_interface.clone(),
+        )
+        .await;
+        device_usage_states.insert(instance_name.to_string(), device_usage_state);
+    }
+
+    let mut usage_ids_to_use = HashMap::<String, (String, String)>::new();
+
+    // Get all available virtual devices, group by device usage status,
+    // a virtual device is available if it's Free or ReservedByConfiguration with requested vdev_ids
+    let mut free_device_usage_states = HashMap::new();
+    let mut reserved_by_configuration_usage_states = HashMap::new();
+    for (instance_name, device_usage_state) in device_usage_states {
+        for (id, state) in device_usage_state {
+            match state {
+                DeviceUsageStatus::Free => {
+                    free_device_usage_states
+                        .entry(instance_name.to_string())
+                        .or_insert(HashSet::new())
+                        .insert(id);
+                }
+                DeviceUsageStatus::ReservedByConfiguration(vdev_id) => {
+                    if requested_vdev_ids.contains(&vdev_id) {
+                        reserved_by_configuration_usage_states
+                            .entry(instance_name.to_string())
+                            .or_insert(HashMap::new())
+                            .insert(id, vdev_id);
+                    }
+                }
+
+                _ => (),
+            };
+        }
+    }
+
+    // Find (Instance, usage slot) for vdev ids, (Instance, usage slot) in free_device_usage_states and
+    // reserved_by_configuration_usage_states are available to be assigned.
+    // Iterate vdev ids to look up vdev id from reserved_by_configuration_usage_states, if not found
+    // pick one (Instance, usage slot) from free_device_usage_states
+    let pending_device_ids = requested_vdev_ids
+        .into_iter()
+        .filter(|vdev_id| {
+            // true means pending
+            let mut resource: Option<(String, String)> = None;
+
+            for (inst, slots) in &reserved_by_configuration_usage_states {
+                for (slot, vid) in slots {
+                    if vdev_id == vid {
+                        resource = Some((inst.clone(), slot.clone()));
+                        break;
+                    }
+                }
+                if resource.is_some() {
+                    break;
+                }
+            }
+
+            if resource.is_none() {
+                if let Some((inst, slots)) = &free_device_usage_states
+                    .iter()
+                    .max_by(|a, b| a.1.len().cmp(&b.1.len()))
+                {
+                    if let Some(slot) = slots.iter().next() {
+                        resource = Some((inst.to_string(), slot.to_string()));
+                    }
+                }
+            }
+
+            match resource {
+                Some((inst, slot)) => {
```

Comment:
nit: use `instance_name` rather than `inst`

Reply:
rename inst to instance_name